### ISSUE TITLE
Unable to connect to remote IPython kernel

### ISSUE THREAD

**Issue by azenin** [spyder-ide/spyder#3323]
## Description

I'm trying to connect to remote IPython kernel

**What steps will reproduce the problem?**

Used this guide https://github.com/ipython/ipython/wiki/Cookbook:-Connecting-to-a-remote-kernel-via-ssh

1. Launched `ipython kernel` on server (Ubuntu 16.04 LTS)
2. Downloaded `kernel-29941.json`
3. Tried to use "Connect to an existing kernel" spyder feature
4. Checked "This is remote kernel"
5. Selected kernel file
6. Pasted some IP address `52.59.__.__`
7. Selected key file `~/.ssh/ipython.pem`

**What is the expected output? What do you see instead?**

Got "Unable to connect to IPython kernel-29941.json"

I've tried console command:
`ipython qtconsole --existing ~/Downloads/kernel-29941.json --ssh 52.59.__.__` - it works

I've also check if the key is ok:
`ssh -i ~/.ssh/ipython.pem 52.59.__.__`

**Please provide any additional information below**

## Version and main components

* Spyder Version: 3.0.0b4 
* Python Version: 2.7.12
* Qt Versions: 4.8.7, PyQt4 (API v2) 4.11.4 on Darwin
* Mac OS

## Dependencies
```
pyflakes >=0.5.0 :  1.2.3 (OK)
pep8 >=0.6       :  1.7.0 (OK)
pygments >=2.0   :  2.1.3 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
nbconvert >=4.0  :  4.2.0 (OK)
pandas >=0.13.1  :  0.18.1 (OK)
numpy >=1.7      :  1.11.1 (OK)
sphinx >=0.6.6   :  1.4.5 (OK)
rope >=0.9.4     :  0.10.3 (OK)
jedi >=0.8.1     :  0.9.0 (OK)
psutil >=0.3     :  4.3.0 (OK)
matplotlib >=1.0 :  1.5.1 (OK)
sympy >=0.7.3    :  1.0 (OK)

```

**Comment by azenin** [comment]
"Unable to connect to IPython kernel" pops up instantly, it seems it didn't even try to connect

**Comment by LLCF** [comment]
I have the same problem

**Comment by azenin** [comment]
The problem is in checking connection file function, to make it work you need to copy connection file to your local `jupyter --runtime-dir`

**Comment by robromanowski** [comment]
@azenin 
Hey Azenin, sorry if this a really obvious answer - where did you get ipython.pem file from? Is that a key you've made yourself?

I'm trying to connect to my remote kernel on my Linux server (RHEL) from Spyder on Windows - and it's failing.

**Comment by azenin** [comment]
@robromanowski yes, you can use `ssh-keygen -t rsa` to generate your own private/public key pair

**Comment by robromanowski** [comment]
@azenin Thanks for the quick response!

That command seems like a great lead. I just ran it, and it seems it made a .pub (instead of .pem) will that work? Or  I do need it to be a .pem

**Comment by azenin** [comment]
@robromanowski try .pub, if it will not work just rename it

**Comment by robromanowski** [comment]
Was gonna try the rename, didn't think it'd be that simple :)

Also, I guess I jumped ahead - I'm having issues finding it.

After i ran `ssh-keygen -t rsa`

It prompted me `Generating public/private rsa key pair.
Enter file in which to save the key (/export/home/rr83008/.ssh/id_rsa): ipython (name i used)
Enter passphrase (empty for no passphrase):
`

And then
`Enter same passphrase again:
Your identification has been saved in ipython.
Your public key has been saved in ipython.pub.
`

Maybe I misunderstand the prompt? I though the location in () the home/username/,ssh/id_rsa was where it was saving it? And I was naming it ipthon? Was I supposed to put a a path there instead of a name?

After it saved I navigated to ~/.ssh, and I don't see the id_rsa directory. Wouldnt it be at id_rsa/KEYNAME.pub?

**Comment by robromanowski** [comment]
@azenin Okay, have taken care of the .pub/.pem issue. I have FTP'd it over to my local Win machine.

I go into Spyder, try to connect now (and using the .pub or renaming it to .pem) neither work. i get the same error you do - immediately.

Although your other comment seems to have solved it with the  --runtime-dir command? Thats where my json is by default? when I do `ipython kernel` and it gives me kernel-#####.json I previously went to that location and FTP'ed the matching number down. I am using that json file in "connect to remore kernel" and the .pub/.pem file I also ftp'd down to my windows machine.

**Comment by azenin** [comment]
you need to copy kernel-#####.json file from _remote_ machine to dir that outputs `jupyter --runtime-dir` command on your _local_ machine

and then you can simply type kernel-##### in spyder menu without any need to locate it

**Comment by robromanowski** [comment]
@azenin Okay, well I was referring to my file was being saved into the -runtime-dir directory on my remote Linux machine...but for my local machine I don't have Juptyer, and besides Spyder lets you navigate to the .json file so would location matter? I get you can type in the exact name without browsing to the file - but wouldnt browsing also work?

Maybe what I'm trying to do won't work.

I have Spyder on my windows machine. I have remote linux machine I have Jupyter working, and I want to be able to use Windows (so the you have a non-laggy GUI, but have it connect to the remote Linux machine so we are in that workspace).

**Comment by robromanowski** [comment]
@ccordoba12 Thanks for the merge!

I did see Paramiko mentioned in one of the many threads I went through. 

Anywhere I can read up on what I need to do with it/set it up?

**Comment by ccordoba12** [comment]
It should work automatically. Have you tried it already?

**Comment by robromanowski** [comment]
I did, but most likely user error on my part. I don't think I fully understand what steps need to be done on Windows vs. Linux.

Paramiko is something I need to run on the Windows side? ( I can set it up with Anaconda?)

I'm trying to make a list of what I need to do on each side:

Linux (on my VM):
1. Find the `juptyer --runtime dir` 
2. Run the `ipython kernel`
3. Know the location of it so I can FTP over to windows

Windows:
1. FTP into the machine, pull off the kernel.json file that matches
2. Open Spyder:
     - use that json
     - use this connection "rob@lab2073" <user>@host
     - attach my .pub (or I can rename it to .pem) ssh that @azenin mentioned above
     - no password for SSH phrase, since I Set it up with no password

When I do that, and try to connect to the linux remote kernel, I get "unable to connect to kernel#####.json

At what point do I do the Paramiko stuff?

**Comment by anuragsodhi** [comment]
Facing the same issue. Mine says unable to open paramiko. I have paramiko installed on both machines

**Comment by Tych0n** [comment]
@anuragsodhi potential solution:

I installed paramiko, checked that it imports in spyder. So i checked versions of IPython on Windows and Linux and they were different:

On Windows 7: python 3.6.1, IPython 5.3.0, Spyder 3.1.4
On Linux: python 3.5.2, IPython 6.2.1

I updated Windows IPython from 5.3.0 to 6.2.1 and spyder remote connection worked! Python versions are still different.

**Comment by RoqueBelda** [comment]
Hello, I have a similar issue, and I have some questions...  I'm trying to configure ipython kernel server on a remote VM Instance in Google Cloud, 

Is it necessary to open the port that use the kernel to connect with the client? If yes, How can I configure the port to avoid that these were randomly configured? 

I have tried to connect several times from Spyder on Windows10 (client) with an SSH key, but I don't get any answers. When I have launched ipthon kernel on the server-side I have assigned the public IP. 

Thanks,

**Comment by seanysull** [comment]
Also having this issue, I get an error like so: 
"Could not open ssh tunnel. The error was:
q must be exactly 160, 224, or 256 bits long"
